Included support for all geocode endpoint parameters

sd opened the pull request:
The existing 'geocode' method only supports the 'responseIncludes' parameter. This pull request expands that support to most additional parameters, as defined in https://github.com/foursquare/twofishes/blob/master/interface/src/main/thrift/geocoder.thrift#L303

sd committed:
Included support for all geocode endpoint parameters

masone commented:
Nice work, thanks!

The `self.geocode` method feels really complex now and I would like to split it up. My first thought was to at least extract the code for each case into a separate method, eg. `handle_radius_params`. There would still be a lot of code around though. So I think I'd probably try to move all of the logic to a separate class that returns the `GeocodeRequest` instance. `self.geocode` would then only have 1-2 lines and you could test the additional logic in isolation.

What do you think?

masone commented on test/twofishes/client_test.rb:
```diff
@@ -29,6 +29,90 @@ def self.geocode(location)
+    it 'should accept max keyword argument' do
+      Twofishes::Client.expects(:thrift_client).returns(ZurichWithParents)
+      result = Twofishes::Client.geocode('Zurich', max: 5).first
+      assert result.parents
```
This test should test that there are max 5 results, but it tests if parents are included

masone commented on test/twofishes/client_test.rb:
```diff
@@ -29,6 +29,90 @@ def self.geocode(location)
+      assert result.parents
+    end
+
+    it 'should accept ll keyword argument' do
```
Please create a separate test for each format for the `ll` argument

masone commented on test/twofishes/client_test.rb:
```diff
@@ -29,6 +29,90 @@ def self.geocode(location)
+    end
+
+    it 'should accept ll keyword argument' do
+      Twofishes::Client.expects(:thrift_client).returns(ZurichWithParents)
```
You can use `Zurich` instead of `ZurichWithParents` in all of your additional tests. This should make it more clear that `ZurichWithParents` is a special version used to test the `ResponseIncludes`

masone commented on test/twofishes/client_test.rb:
```diff
@@ -29,6 +29,90 @@ def self.geocode(location)
+    it 'should accept ll keyword argument' do
+      Twofishes::Client.expects(:thrift_client).returns(ZurichWithParents)
+      result = Twofishes::Client.geocode('Zurich', ll: "47,8").first
+      assert result.displayName
```
For consistencies sake, I would do the same assertion as in the first test, see https://github.com/arepalabs/twofishes-ruby/blob/include-all-geocode-parameters/test/twofishes/client_test.rb#L23

masone commented on test/twofishes/client_test.rb:
```diff
@@ -29,6 +29,90 @@ def self.geocode(location)
+    it 'should accept radius keyword argument' do
+      Twofishes::Client.expects(:thrift_client).returns(ZurichWithParents)
+      result = Twofishes::Client.geocode('Zurich', radius: 5000).first
+      assert result.parents
```
This test should test that the radius attribute works, but it just tests if it returns results.

masone commented on test/twofishes/client_test.rb:
```diff
@@ -29,6 +29,90 @@ def self.geocode(location)
+      assert result.parents
+    end
+
+    it 'should accept bounds keyword argument' do
```
Please create a separate test for each format of the `bounds` argument.

masone commented:
I left some comments on the tests. There are a couple of recurring points I'd like you to have a look at. Let me know if I can help.

sd committed:
Refactored request parameter handling into a separate class. 

Now both geocode and reverse_geocode can accept all possible parameters, along with a few helper transformations

sd commented:
Take a look at these changes... refactored all request processing into a separate class, and tested it extensively.

Used that class for both geocode and reverse geocode, and tweaked client_test to address the new behaviour. And while doing that, changed a bit the mocking of the response so we could test the request being passed correctly.

masone commented:
Awesome, thank you! It feels way cleaner this way, I'm sure codeclimate will like that too. I'll have to take a deeper look at it. And I'll have to make some decisions on how to proceed with some inherit changes.

sd committed:
pry out pry

masone commented on lib/twofishes/geocode_request.rb:
```diff
@@ -0,0 +1,57 @@
+      when Array
+        GeocodePoint.new(lat: ll[0].to_f, lng: ll[1].to_f)
+      when Hash
+        GeocodePoint.new(lat: (ll[:lat] || ll['lat']).to_f, lng: (ll[:lng] || ll['lng'] || ll[:lon] || ll['lon']).to_f)
```
Should we move handling the aliases to `substitute_aliases`? Also I would leave out the string keys. Same for bounds below.

masone commented on lib/twofishes/geocode_request.rb:
```diff
@@ -0,0 +1,57 @@
+    end
+
+    def substitute_aliases(options)
+      options[:maxInterpretations] = options.delete(:max) || options.delete(:max_interpretations)
```
Shall we assign with `||=` in order to make sure the most exact option does not get overwritten? Eg. if I would have `{maxInterpretations: 5, max: 999}` it would take 5 and not overwrite it by 999?

masone commented on test/twofishes/client_test.rb:
```diff
@@ -3,35 +3,81 @@
   class Zurich
     def self.geocode(location)
-      YAML::load_file('test/fixtures/response.yaml')
+      if location.responseIncludes == [ResponseIncludes::PARENTS]
```
This feels really weird, especially now that it has so much logic in it. Maybe I should re-think the way we test this.

sd replied:
I agree it feels overly complicated... but I felt we should test not only that 'geocode' was called, but that the relevant parameters were passed and reflected in the response.

masone replied:
I agree. But I fear we're testing the fake geocode method instead of the real one.

masone commented:
Sebastian, I now had time to properly review your changes. I like it a lot!
You're right, I now think too we should get away with the change to options. The aliases are a nice addon too. I left 3 more small comments. If you're still motivated to work on it, go for it. Let me know otherwise and I will merge and do the small changes outside this PR.

sd committed:
simpler handling of lat/lng (or lon) parameters

sd committed:
prioritize most specific parameter alias

sd committed:
fixed lat/lng option processing

masone commented:
Wow, thank you so much for your effort. This adds a lot of value to the gem, love it!

masone commented:
Published [2.1.0.pre](https://rubygems.org/gems/twofishes/versions/2.1.0.pre.rc.1)
